### ISSUE TITLE
[wx] Modernization of Change Password Dialog

### ISSUE THREAD

**Issue by rafaelx** [pwsafe/pwsafe#1226]
This ticket/issue was created to track and discuss the changes to the password change dialog.
![ChangePasswordDialog](https://github.com/pwsafe/pwsafe/assets/4042917/c45cd7c4-e554-411a-aaa4-95cee3dd21f9)
![ChangePasswordDialogYubiInserted](https://github.com/pwsafe/pwsafe/assets/4042917/a9b663d2-e247-494d-afd7-2ce09ce8a95b)
Related to #1197.

**Comment by igor2x** [comment]
1. Looking at text: "Enter the current master password followed by a new one. Type the new one again to confirm it." it looks to me redundant. From three fields design isn't it obvious all three fields should be filled in? In my humble opinion, this text doesn't tell anything that is not really obvious.
I see two options:
a) Remove above text and that is (if we want to go ultra simplistic).
b) But... in my humble opinion, because this is potentially dangerous action and if new master password is not remembered it can lead to critically bad consequences, there should be some text, but explaining what will this change do "Setting new master password and it will be from now on used to unlock Password Safe database." Maybe something similar, I am not a native English speaker, so text probably can be improved.

2. Obviously "Show Master Password" checkbox should be removed, because now we have eye icons. I checked what does Windows use when pressed Alt+Ctrl+Delete to change password. Windows use eye icon at each of fields, just like we have. Selecting "Show Master Password" checkbox it displays all passwords, that may be convenient, but on the other hand how frequent do we change master password? Not really frequent, so having eye icons are fine in my opinion.

3. OK button should be replace with: "Replace" or probably even "Change".

4. Remove icons on buttons.

---

For second Yubikey dialog: I don't have Yubikey... I don't want to comment too much, because my assumptions are probably wrong. Just wondering how to get rid of text bellow or how to simplify the text. Specially "top Yubikey" is what I don't like. Why is bottom Yubikey not greyed out if it should not be clicked?

**Comment by nobugshere** [comment]
I just found a couple of bugs in this dialog.  If the safe only uses a Yubikey (no password):
1. After activating the key as usual, it grays-out the second YK button, expecting you to insert a different key, so you cannot add a PW and keep the same YK.  You can accomplish this, but it takes two passes.  First, change it to PW only, then a second invocation to add the YK back.
2. If I want to remove the YK and just use a PW, I can fill-in the new and confirmation boxes, but when I click OK, it says "The combination cannot be blank."  I have to put *something* in the Old Master Password field; anything at all works.

**Comment by igor2x** [comment]
@nobugshere, are those issues new in Password Safe v1.19-pre on master branch or are they already present on officially stable v1.18.2 version?

**Comment by nobugshere** [comment]
@igor2x,
Yes, I'll open a new issue.  It was your mention of the Yubikey buttons that made me look at them :-)

**Comment by igor2x** [comment]
@nobugshere, can you please now remove those two replies from this issue and I will remove my two, to clean this issue up.

**Comment by rafaelx** [comment]
A first interim result.

![ChangePasswordDialogNew](https://github.com/pwsafe/pwsafe/assets/4042917/f54ec558-370b-483d-8065-05297d2b5fdb)
![ChangePasswordDialogNewYubiInserted](https://github.com/pwsafe/pwsafe/assets/4042917/3120db42-ce67-4503-8d49-c8de436747e3)

The completion of the work on the dialog awaits the acceptance of #1232.

**Comment by rafaelx** [comment]
By introducing the eye icon for the password input fields, the showing and hiding of the password can be controlled separately for each input field. In the current version of the password change dialog, there is a single checkbox for this. This can be used to show the entered old password and the new password. If the old and new passwords are shown, the entry in the password confirmation field is deleted and this is deactivated.

The status of the checkbox also affects the verification routine that is triggered by the OK button. If the old password and the new one are hidden, the new password and the confirmation password are compared. If, on the other hand, the old password and the new password are shown, only the strength of the new password is checked.

Do we still need this functionality and how should it behave in combination with the functionality of the password input fields with eye icon?

The code that affects the password input fields and also sets m_isPasswordHidden for the verification code.
```
  auto showCombinationCheckBox = new wxCheckBox(this, ID_SHOWCOMBINATION, _("Show Master Password"), wxDefaultPosition, wxDefaultSize, 0 );
  showCombinationCheckBox->SetValue(false);
  m_isPasswordHidden = true;
  showCombinationCheckBox->Bind(wxEVT_CHECKBOX, [&](wxCommandEvent& event) {
    m_oldPasswdEntry->SecureTextfield(!event.IsChecked());
    m_newPasswdEntry->SecureTextfield(!event.IsChecked());
    m_confirmEntry->DisableAndClear(event.IsChecked());
    m_isPasswordHidden = !event.IsChecked();
    if(event.IsChecked())
      m_confirmEntry->AllowEmptyCombinationOnce();
    else
      m_confirmEntry->ForbidEmptyCombinationOnce();
  });
```

The verification code.
```
void SafeCombinationChangeDlg::OnOkClick(wxCommandEvent& WXUNUSED(evt))
{
  if (Validate() && TransferDataFromWindow()) {
    StringX errmess;
    const StringX old = m_oldresponse.empty() ? m_oldpasswd : m_oldresponse;
    int rc = m_core.CheckPasskey(m_core.GetCurFile(), old);
    if (rc == PWScore::WRONG_PASSWORD) {
      wxMessageDialog err(this, _("The old master password is not correct"),
                          _("Error"), wxOK | wxICON_EXCLAMATION);
      err.ShowModal();
    } else if (rc == PWScore::CANT_OPEN_FILE) {
      wxMessageDialog err(this, _("Cannot verify old master password - file gone?"),
                          _("Error"), wxOK | wxICON_EXCLAMATION);
      err.ShowModal();
    } else if (m_isPasswordHidden && (m_confirm != m_newpasswd)) {
      wxMessageDialog err(this, _("New master password and confirmation do not match"),
                          _("Error"), wxOK | wxICON_EXCLAMATION);
      err.ShowModal();
    // Vox populi vox dei - folks want the ability to use a weak
    // passphrase, best we can do is warn them...
    // If someone want to build a version that insists on proper
    // passphrases, then just define the preprocessor macro
    // PWS_FORCE_STRONG_PASSPHRASE in the build properties/Makefile
    // (also used in CPasskeySetup)
    } else if (!CPasswordCharPool::CheckMasterPassword(m_newpasswd, errmess)) {
      wxString msg = errmess.c_str();
#ifndef PWS_FORCE_STRONG_PASSPHRASE
      msg += wxT("\n");
      msg += _("Use it anyway?");
      wxMessageDialog err(this, msg,
                          _("Weak Master Password"), wxYES_NO | wxNO_DEFAULT | wxICON_EXCLAMATION);
      err.SetYesNoLabels(_("Use anyway"), _("Cancel"));
      int rc1 = err.ShowModal();
      if (rc1 == wxID_YES)
        EndModal(wxID_OK);
#else
      wxMessageDialog err(this, msg,
                          _("Error"), wxOK | wxICON_HAND);
      err.ShowModal();
#endif // PWS_FORCE_STRONG_PASSPHRASE
    } else { // password checks out OK.
      EndModal(wxID_OK);
    }
  }
}
```

**Comment by nobugshere** [comment]
Personally, given that the reason for having two fields is to check what the user typed, I think the comparison should be unconditional.  I would suggest removing all of the isPasswordHidden stuff.  With a separate eye button for each text field, each button should only affect it's associated text field and should not have any other affect on the program when the user clicks OK.  Any checks performed should be independent of the visibility state of the text.

With that in mind, I was just looking at this code because I noticed that the OnYubiBtn2Click() function does not compare the two fields at all!  Even when using a Yubikey, the "new" and "confirm" entries should be compared.  (I noticed the strength check is not being done, but it's not done in the "Set Master Password" dialog either.  I'm not sure if it's intended to be that way with a Yuibkey.)  Last night I started to work on a fix for this and #1230, then I noticed you were still working on the dialog update.  If you don't fix it, I'll take it up again when you're finished.
